### PR TITLE
GTC Optimizations Part 1: Basic Horizontal Execution Merging and Temporary Replacement

### PR DESCRIPTION
**Pull request by fthaler**
## Description

This introduces basic (greedy) horizontal execution merging and replacement of temporary fields by local scalars as optimization passes on the OIR level. Local scalars are introduced in the OIR and GTCPP IRs.

## Requirements

Before submitting this PR, please make sure:

- [x] The code builds cleanly without new errors or warnings
- [x] The code passes all the existing tests
- [x] If this PR adds a new feature, new tests have been added to test these
new features
- [x] All relevant documentation has been updated or added

Additionally, if this PR contains code authored by new contributors:

- [x] All the authors are covered by a valid contributor assignment agreement,
signed by the employer if needed, provided to ETH Zurich
- [x] The names of all the new contributors have been added to an updated
version of the AUTHORS.rst file included in the PR

### CODE REVIEW

File: src/gtc/gtcpp/gtcpp.py [hunk 1]
```diff
@@ -246,12 +241,6 @@ class GTStage(LocNode):
     # or `temporaries`
     args: List[Arg]
 
-    @validator("args")
```
Comment (havogt) [inline src/gtc/gtcpp/gtcpp.py:249]: Why did you remove it? I think if this is removed we should add a test to `test_gtcpp_compilation.py` that ensures a stage with no arguments produces compilable code.

File: src/gtc/oir.py [hunk 2]
```diff
@@ -120,13 +120,18 @@ class ScalarDecl(Decl):
     pass
 
 
+class LocalScalar(ScalarDecl):
+    pass
+
+
 class Temporary(FieldDecl):
```
Comment (havogt) [inline src/gtc/oir.py:127]:
```suggestion
class Temporary(Decl):
```
because otherwise we can pass a temporary as param to the stencil

File: src/gtc/oir.py [hunk 3]
```diff
@@ -120,13 +120,18 @@ class ScalarDecl(Decl):
...
     pass
 
 
 class HorizontalExecution(LocNode):
     body: List[Stmt]
     mask: Optional[Expr]
+    declarations: List[LocalScalar]
 
     @validator("mask")
     def mask_is_boolean_field_expr(cls, v: Optional[Expr]) -> Optional[Expr]:
```
Comment (havogt) [inline src/gtc/oir.py:137]: Please change in line 153

```
params: List[Decl]
```
to
```
params: List[Union[ScalarDecl,FieldDecl]]
```
and fix the test `params` is only for parameters of the stencil. Temporaries should be added to a `declarations` node if that is needed on the Stencil level.

File: src/gtc/oir.py [hunk 4]
```diff
@@ -120,13 +120,18 @@ class ScalarDecl(Decl):
     pass
 
 
+class LocalScalar(ScalarDecl):
```
Comment (havogt) [inline src/gtc/oir.py:123]: Is it needed, se also my other suggestions below

**Review by havogt — CHANGES_REQUESTED** [review]

File: src/gtc/gtcpp/gtcpp.py (hunk 1 above)
Comment (fthaler) [reply src/gtc/gtcpp/gtcpp.py:249]: Because there is _test_stage_without_effect_ that assigns to a temporary and does nothing else. After the temporary is replaced by a scalar in the optimization passes, there are no arguments left. And it works, so this test basically exists.

File: src/gtc/oir.py (hunk 3 above)
Comment (fthaler) [reply src/gtc/oir.py:137]: Can we change that in a separate PR? That has nothing to do with the optimizations introduced here, this all works the same way as it has before.

File: src/gtc/oir.py (hunk 4 above)
Comment (fthaler) [reply src/gtc/oir.py:123]: Can not see any other suggestions below…

File: tests/test_unittest/test_gtc/test_passes/test_oir_optimizations/test_horizontal_execution_merging.py [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+        .build()
+    )
+    transformed = merger.visit(testee)
+    if isinstance(merger, ZeroOffsetMerging):
+        assert transformed == testee
+    else:
+        assert len(transformed.horizontal_executions) == 1
+        assert transformed.horizontal_executions[0].body == sum(
+            (he.body for he in testee.horizontal_executions), []
+        )
```
[the remark this answers is not in the document]
Comment (fthaler) [reply tests/test_unittest/test_gtc/test_passes/test_oir_optimizations/test_horizontal_execution_merging.py]: I added a self-assignment (without offset) to the test, seems to work fine.

[the remark this answers is not in the document]
Comment (havogt) [reply tests/test_unittest/test_gtc/test_passes/test_oir_optimizations/test_horizontal_execution_merging.py]: Sorry that was a pending comment from Monday.

File: src/gtc/oir.py (hunk 3 above)
Comment (fthaler) [reply src/gtc/oir.py:137]: Currently difficult due to pydantic Union quirks.

Comment (havogt) [reply src/gtc/oir.py:137]: I would like to fix it here, because the tests add a temporary to params which is wrong.
Since we anyway, want to move away from pydantic, we can do a temporary hacky check, like

```
class FieldDecl(Decl):
    field_decl=""
```
etc. then the Decls are incompatible with each other.

File: src/gtc/oir.py (hunk 4 above)
Comment (havogt) [reply src/gtc/oir.py:123]: Below in the file, i.e. the previous comment about fixing params. I meant, is there a good reason to distinguish ScalarDecl as parameter and LocalScalar as a declaration inside the body.

File: src/gtc/gtcpp/gtcpp.py (hunk 1 above)
Comment (havogt) [reply src/gtc/gtcpp/gtcpp.py:249]: Can you still add an explicit test.

File: src/gtc/oir.py (hunk 3 above)
Comment (fthaler) [reply src/gtc/oir.py:137]: Does not help at all as long as temporary has the same fields as Decl I guess. The only fix that works is deriving the `Temporary` directly from `Decl` and adding an extra field to it. But as this probably only fixes this specific issue, but not the same issue the other way around nor the issue in general, I vote for postponing this until pydantic is replaced and this works without workarounds.

Comment (havogt) [reply src/gtc/oir.py:137]: yes, that's another comment I made above or below

File: src/gtc/oir.py (hunk 4 above)
Comment (fthaler) [reply src/gtc/oir.py:123]: I guess so. First, they are declared at a different point, ScalarDecl are global parameters, passed as stencil arguments, LocalScalars are local to a horizontal execution. Also, they mean very different things, the ScalarDecl stands for one scalar for the whole domain, LocalScalar stands for a different scalar at each grid point. Further, the lowering to GTCPP is simpler (and maybe other possible IRs) as the distinction has to be handled there, too.

File: src/gtc/oir.py (hunk 3 above)
Comment (havogt) [reply src/gtc/oir.py:137]: All I care for now is not adding a temporary to `params` and if not in this PR, then in another PR, directly afterwards fix the grammar even if pydantic cannot enforce it.

File: src/gtc/oir.py (hunk 4 above)
Comment (havogt) [reply src/gtc/oir.py:123]: But then it shouldn't inherit from ScalarDecl?

File: src/gtc/oir.py (hunk 3 above)
Comment (fthaler) [reply src/gtc/oir.py:137]: The problem is that pydantic does not only not enforce it, but helps destroying it by its invalid conversions…

File: src/gtc/oir.py (hunk 4 above)
Comment (fthaler) [reply src/gtc/oir.py]: True, fixed!

File: src/gtc/oir.py (hunk 3 above)
Comment (fthaler) [reply src/gtc/oir.py:137]: I added a comment.

File: src/gtc/oir.py (hunk 2 above)
Comment (fthaler) [reply src/gtc/oir.py]: Done.

**Review by havogt — APPROVED** [review]

File: src/gtc/gtcpp/gtcpp.py (hunk 1 above)
Comment (fthaler) [reply src/gtc/gtcpp/gtcpp.py:249]: Done.

**Review by havogt — APPROVED** [review]

**Comment by DropD** [comment]
bors try

**Comment by havogt** [comment]
bors try

**Comment by havogt** [comment]
bors try

**Comment by havogt** [comment]
Gitlab tests passed! I'll merge now. Then master should be green again.
